src/components/structures/MatrixChat.js

```diff
@@ -1304,9 +1304,20 @@ export default React.createClass({
             }
         });
 
-        const dft = new DecryptionFailureTracker((total) => {
-            // TODO: Pass reason for failure as third argument to trackEvent
-            Analytics.trackEvent('E2E', 'Decryption failure', 'unspecified_error', total);
+        const dft = new DecryptionFailureTracker((total, errorCode) => {
+            Analytics.trackEvent('E2E', 'Decryption failure', errorCode, total);
+        }, (errorCode) => {
+            // Map JS-SDK error codes to tracker codes for aggregation
+            switch (errorCode) {
+                case 'MEGOLM_UNKNOWN_INBOUND_SESSION_ID':
+                    return 'olm_keys_not_sent_error';
+                case 'OLM_UNKNOWN_MESSAGE_INDEX':
+                    return 'olm_index_error';
+                case undefined:
+                    return 'unexpected_error';
+                default:
+                    return 'unspecified_error';
```

Comment:
Is there a reason to map these to a different set of strings rather than use what the js-sdk emits?

Reply:
Separate the implementation of the js-sdk from the stats in our analytics.